Added support for `revert_to_preemptible` and `optimization_window` under `strategy` object.

sharadkesarwani opened the pull request:
Added support for RevertToPreemptible and OptimizationWindows under strategy object.

# Jira Ticket
Ref: https://spotinst.atlassian.net/browse/SPOTAUT-19474

sharadkesarwani committed:
Added support for `revert_to_preemptible` and `optimization_window` under strategy object

sharadkesarwani committed:
upgraded SDK version

sharadkesarwani committed:
updated changelog

chandra1-n commented on docs/resources/elastigroup_gcp.md:
```diff
@@ -27,6 +27,10 @@ resource "spotinst_elastigroup_gcp" "example" {
+  revert_to_preemptible{
+    perform_at="timeWindow"
+  }
+  optimization_window=["Mon:01:00-Mon:03:00"]
```
change the name to optimization_windows

sharadkesarwani replied:
updated

chandra1-n commented on docs/resources/elastigroup_gcp.md:
```diff
@@ -141,7 +145,9 @@ The following arguments are supported:
-
+* `revert_to_preemptible` - (Optional) Setting for revert to preemptible option.
+  * `perform_at` - (Required) Valid values: "always", "never", "timeWindow". Required on strategy.revertToPreemptible object.
+* `optimization_window` - (Optional)(Array of strings) Set time window to perform the revert to preemptible. Time windows must be at least 120 minutes. Format: DayInWeek:HH-DayInWeek:HH. Required when strategy.revertToPreemptible.performAt is 'timeWindow'.
```
remove array of strings and change the name to optimization_windows

sharadkesarwani replied:
updated

chandra1-n commented on go.sum:
```diff
@@ -541,6 +541,8 @@ github.com/spaolacci/murmur3 v0.0.0-20180118202830-f09979ecbc72/go.mod h1:JwIasO
 github.com/spf13/pflag v1.0.2/go.mod h1:DYY7MBk1bdzusC3SYhjObp+wFpr4gzcvqqNjLnInEg4=
 github.com/spotinst/spotinst-sdk-go v1.362.0 h1:hyhySd+WTZfC/J5670aumE+oWejGRRQ0ay8WF01+Pr4=
```
remove 362 version packages

sharadkesarwani replied:
removed

chandra1-n commented on spotinst/elastigroup_gcp_strategy/consts.go:
```diff
@@ -8,4 +8,7 @@ const (
 	DrainingTimeout       commons.FieldName = "draining_timeout"
 	FallbackToOnDemand    commons.FieldName = "fallback_to_ondemand"
 	ProvisioningModel     commons.FieldName = "provisioning_model"
+	OptimizationWindow    commons.FieldName = "optimization_window"
```
change the name to optimization_windows

sharadkesarwani replied:
updated

chandra1-n commented on spotinst/elastigroup_gcp_strategy/fields_spotinst_elastigroup_gcp_strategy.go:
```diff
@@ -230,4 +231,139 @@ func Setup(fieldsMap map[commons.FieldName]*commons.GenericField) {
+	)
+}
+func flattenRevertToPreemptible(revertToPreemptible *gcp.RevertToPreemptible) []interface{} {
+	result := make([]interface{}, 0, 1)
```
why hardcoded to 0, 1

sharadkesarwani replied:
Because revertToPreemptible is an pointer to object and not an array so we cant use len() to create list of interface{}.

chandra1-n reviewed (changes requested):
please fix the review comments

sharadkesarwani committed:
updated `optimization_window` to `optimization_windows` as part of review comment

updated `optimization_window` to `optimization_windows` as part of review comment

sharadkesarwani committed:
executed "go mod tidy"
